daffodil-lib/src/main/scala/org/apache/daffodil/util/Delay.scala

```diff
@@ -27,18 +27,71 @@ package org.apache.daffodil.util
  * ones we might have wanted to supply later, we instead supply them, but
  * as Delay objects.
  *
- * For more info, this is used in the IncludeImport stuff in DSOM.
+ * For more info, this is used in the IncludeImport stuff in DSOM, and
+ * in numerous other places where we construct, functionally, cyclic graphs.
  */
+final class Delay[T] private (private var box: Delay.Box[T])
+  extends PreSerialization {
+  //
+  // This trick of taking another object on a var, and
+  // then clobbering the var after we demand the value
+  // eliminates holding onto a bunch of objects due
+  // to closures for the pass-by-name arguments.
+  //
+  // The idea is no matter what the Scala implementation is doing in its implementation of
+  // by-name args, that's on the constructor of the box object,
+  // and we're going to explicitly null-out the reference to the box object
+  // which guarantees we're not holding onto things we don't expect
+  // once the Delay object has been evaluated.
+  //
+  lazy val value: T = {
+    val v = box.value
+    box = null // throws away box, which allows GC of all closures, etc.
+    // Delay.evalDelay()
+    v
+  }
+
+  /**
+   * For creating a delay object purely to satisfy a type requirement
+   * when you know the argument does not actually need to be delayed.
+   */
+  def force : Delay[T] = { value; this }
 
-class Delay[T] private (v: => T) {
-  lazy val value = v
-  private val hasValue: Boolean = false
+  /**
+   * Create a string representation. Does not force the value to be computed.
+   */
   override def toString = {
-    val bodyString = if (hasValue) value.toString else "..."
+    val bodyString = if (box eq null) value.toString else "..."
     "Delay(" + bodyString + ")"
   }
+
+  override def preSerialization = {
+    value // force evaluation
```

Comment:
I wonder if this really wants be an ``Assert(box != null)``. I imagine if we are serilazing a Delay, it's because it's part of the stuff needed for runtime, and that fact that it wasn't evaluated means that there was a bug in the code and we a missing something that forces that evaluation?

Reply:
I will give this a try and test it. I believe there are still some delays that aren't being forced because they are created in places outside of the OOLAG framework where errors can be caught properly.

Reply:
This is now an assertion check that the Delay must have value or it fails.